clib/astlib.c

```diff
@@ -0,0 +1,200 @@
+#include "astlib.h"
+
+/*
+Mutates AST according to action taken
+
+Input:
+    - ast: struct representing the AST
+    - action: index of action
+
+Mutates:
+    - ast->nodes, ast->edges
+*/
+void take_action(State *ast, int action)
+{
+    change_zast(action);
+    get_ast();
+    get_cursor_info();
+    copy_ast(ast, &curr_state);
+}
+
+/*
+Test the code given by the AST on the unit test
+
+Input:
+    - ast: struct representing the AST
+    - unit_test_index: index of unit test
+
+Output:
+    - 0 for not passed; 1 for passed
+*/
+int check_ast(State *ast)
+{
+    copy_ast(&curr_state, ast);
+    return run_unit_tests();
+}
+
+/*
+Find the valid actions given the AST
+
+Input:
+    - ast: struct representing the AST
+
+Mutates:
+    - ast->permitted_actions
+*/
+void valid_actions(State *ast)
+{
+    for (int i = 0; i < NUM_ACTIONS; i++)
+        ast->permitted_actions[i] = 1;
+
+    curr_state = *ast;
```

Comment:
This is a dead function. Its functionality is done by `get_cursor_info` since the cursor info provides what actions can be performed. The function is deleted.